### PR TITLE
PERF: use isinstance(obj, Foo) instead of ABCFoo

### PR DESCRIPTION
**Pull request by jbrockmendel**
per discussion in #27353, isinstance checks with the ABCFoo are about 4x as expensive as the Foo checks.  Individually they are still very small, but they add up.

### CODE REVIEW

**Review by jorisvandenbossche — APPROVED** [review]

File: pandas/core/frame.py [hunk 1]
```diff
@@ -7020,8 +7014,8 @@ def _gotitem(
         self,
         key: Union[str, List[str]],
         ndim: int,
-        subset: Optional[Union[Series, ABCDataFrame]] = None,
-    ) -> Union[Series, ABCDataFrame]:
+        subset: Optional[Union[Series, "DataFrame"]] = None,
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:7017]: _typing.FrameOrSeriesUnion or _typing.FrameOrSeries?

File: pandas/core/frame.py [hunk 2]
```diff
@@ -7020,8 +7014,8 @@ def _gotitem(
...
+    ) -> Union[Series, "DataFrame"]:
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:7018]: same

**Review by simonjayhawkins — APPROVED** [review]
Thanks @jbrockmendel lgtm ex use of aliases

**Comment by jorisvandenbossche** [comment]
Thanks!

**Pull request merged**